### ISSUE TITLE
Testcafe throws 'Cannot read properties of null' error and hangs for all PayPal Scenarios on latest 1.18 Version

### ISSUE THREAD

**Issue by royalkanai** [DevExpress/testcafe#6918]
### What is your Scenario?

The use-case is to navigate to promotional petition and contribute the amount using paypal. Once the payment is confirmed. site redirects to the thankyou page.  

### What is the Current behavior?

Once the payment has been made through paypal, it hangs and unable to validate the thankyou page. 

### What is the Expected behavior?

Expected behaviour would be TestCafe to run without interruption and validate the thankyou page. 

### What is your public website URL? (or attach your complete example)

Change Staging: https://www.staging-change.org/

### What is your TestCafe test code?
**Note:** Please feel free to remove wait statements if test passes. 
```js
/* global window */
/* eslint-disable no-await-in-loop, no-param-reassign */
import { Chance } from 'chance';
import { ClientFunction, RequestLogger, Selector, t } from 'testcafe';
import logToConsole from '../support/custom_commands/logToConsole';
import { httpAuthCredentials } from '../support/helpers';

const chance = new Chance();
const emailAddress = chance.email();
const paypalIframe = Selector('[data-testid="paypal-button"] iframe').withAttribute('src');
const ppmuPaypalIframe = Selector('#modal-paypal-button iframe').withAttribute('src');
const payWithPaypalButton = Selector('[data-funding-source="paypal"]').filterVisible();

// PayPal Window
const emailInput = Selector('#email');
const passwordInput = Selector('#password');
const nextButton = Selector('#btnNext');
const loginButton = Selector('#btnLogin');
const consentButton = Selector('#consentButton, [data-test-id="continueButton"]').filterVisible();
const submitButtonInitial = Selector('[data-testid="submit-button-initial"]');

const chargeLogger = RequestLogger(
  { url: 'https://www.staging-change.org/api-proxy/-/payments/charges', method: 'post' },
  {
    logRequestHeaders: false,
    logRequestBody: true,
    stringifyRequestBody: true,
  }
);

const submitWithPaypalButton = async (ppmu = false) => {
  const paypalIframeSelector = ppmu ? ppmuPaypalIframe : paypalIframe;
  await t
    .switchToIframe(paypalIframeSelector)
    .hover(payWithPaypalButton)
    .click(payWithPaypalButton)
    .wait(500);
};

const paypalLogin = async () => {
  await t
    .typeText(emailInput, '', { timeout: 30000 })
    .click(nextButton)
    .typeText(passwordInput, '')
    .click(loginButton);
};

const completePaypalMiniWindow = async ({ shouldLogin = true, oneTimePayment = false } = {}) => {
  if (shouldLogin) {
    await paypalLogin();
  }

  const continueButton = oneTimePayment ? submitButtonInitial : consentButton;
  // const paymentOptions = Selector('#paymentMethod, [data-testid="stacked-payment-options"]').filterVisible();

  await t.hover(continueButton).click(continueButton);
};

const checkResponseExists = async (logger, maxRetry = 10) => {
  logToConsole('Check Response');
  while (logger.requests.find(r => !r.response) && maxRetry > 0) {
    maxRetry -= 1;
    await t.wait(500);
  }
  return logger.requests.find(r => r.response);
};

fixture('Verify PayPal Payments')
  .page('https://www.staging-change.org/')
  .requestHooks(chargeLogger)
  .httpAuth(httpAuthCredentials);

test.skip('Create Membership with Paypal', async () => {
  // Navigate to the Membership page, select contribution amount and pay with PayPal
  await t
    .navigateTo('/s/member')
    .click(Selector('[data-testid="member_landing_inline_plan_button_wrapper"]').nth(0))
    .expect(Selector('[data-testid="member_payment_form"]').visible)
    .ok()
    .click(Selector('[data-testid="payment-option-button-paypal"]'));

  await t
    .expect(Selector('[data-testid="input_email"]').visible)
    .ok()
    .typeText(Selector('[data-testid="input_email"]').filterVisible(), emailAddress)
    .typeText(Selector('[data-testid="input_confirmation_email"]').filterVisible(), emailAddress)
    .typeText(Selector('[data-testid="input_first_name"]'), 'first')
    .typeText(Selector('[data-testid="input_last_name"]'), 'last');

  await t
    .wait(500)
    .switchToIframe(paypalIframe)
    .hover(payWithPaypalButton)
    .click(payWithPaypalButton)
    .wait(500);

  // PayPal Window
  await t
    .typeText(emailInput, '')
    .click(nextButton)
    .typeText(passwordInput, '')
    .click(loginButton);

  await t
    .expect(Selector('#paymentMethod, [data-testid="stacked-payment-options"]').filterVisible().visible)
    .ok()
    .hover(consentButton)
    .click(consentButton);

  // Main page
  await t.expect(Selector('[data-view="subscription_plans/thank_you"]').visible).ok();

  const getCurrentUrl = ClientFunction(() => window.location.href);
  await t.expect(await getCurrentUrl()).contains('/s/member/thank_you');
});

test('Promote with PayPal', async () => {
  await t
    .navigateTo('/p/asdf-stgaqaoppscore8/sponsors/new')
    .expect(Selector('.js-payment-form').visible)
    .ok()
    .typeText(Selector('[data-testid="payment-form-amount-input"]'), '30', { replace: true })
    .click(Selector('[data-testid="payment-option-button-paypal"]'));

  await t
    .expect(Selector('[data-testid="input_email"]').visible)
    .ok()
    .typeText(Selector('[data-testid="input_email"]').filterVisible(), emailAddress)
    .typeText(Selector('[data-testid="input_confirmation_email"]').filterVisible(), emailAddress)
    .typeText(Selector('[data-testid="input_first_name"]'), 'first')
    .typeText(Selector('[data-testid="input_last_name"]'), 'last');
  await submitWithPaypalButton();

  // PayPal Window
  await completePaypalMiniWindow();

  // XHR Request
  await t
    .expect(chargeLogger.count(() => true))
    .gt(0, { timeout: 30000 })
    .expect(await checkResponseExists(chargeLogger))
    .ok({ timeout: 30000 });

  logToConsole('Return to main page');
  // Main page
  await t
    .expect(Selector('[data-testid="ppmu-modal"]').exists)
    .ok()
    .click(Selector('[data-testid="ppmu-no"]'));
    logToConsole('Decline PPMU');
  await t.expect(Selector('[data-testid=promote-thank-you-views-text]').visible).ok();
});
```

### Your complete test report
 1) TypeError: Cannot read properties of null (reading '0')
      Browser: Chrome 99.0.4844.51 / macOS 10.15.7  

         270 |        this._handlePlainObject(val);
         271 |};
         272 |
         273 |DecodingTransformer.prototype.transform = function () {
         274 |    this.visitedRefs[0] = true;
         275 |  >  this._handleValue(this.references[0], this.references, 0);
         276 |
         277 |    return this.references[0];
         278 |};
         279 |
         280 |

         at DecodingTransformer.transform (/Users/kshah/work/kanai/Testcafe18/regressionqaa/node_modules/replicator/index.js:275:38)
         at module.exports.Replicator.decode (/Users/kshah/work/kanai/Testcafe18/regressionqaa/node_modules/replicator/index.js:577:24)

_No response_

### Steps to Reproduce

1. Take the sample TestCafe code attached that has all steps to reproduce 
2. Run the test against a latest TestCafe version 
3. Observe the behaviour 

### TestCafe version

1.18.5-alpha.1

### Command-line arguments

testcafe chrome paypal-demo.js

### Browser name(s) and version(s)

Chrome 99.0.4844.51

### Platform(s) and version(s)

macOS 10.15.7

**Comment by AlexKamaev** [comment]
I was not able to reproduce the issue using your example, since it uses HTTP authentication. Please share credentials to your website with us to reproduce the issue on our side. If you cannot share the credentials here, please send them at [support@devexpress.com](mailto:support@devexpress.com)

I also tried to reproduce the issue using the following PayPal demo page:
[https://demo.paypal.com/us/demo/go_platform/pcRestServerV2/product](https://demo.paypal.com/us/demo/go_platform/pcRestServerV2/product)

For this purpose I created a sandbox account as described here:
[https://developer.paypal.com/developer/accounts/](https://developer.paypal.com/developer/accounts/)

This is the code of the test I tried:


```JS
fixture `f`
    .page `https://demo.paypal.com/us/demo/go_platform/pcRestServerV2/product`;


test('t', async t => {
    await t.switchToIframe('iframe[title=PayPal]');
    await t.click('.paypal-button')
    await t.typeText('#email', sandbox_user_email);
    await t.click('#btnNext');
    await t.typeText('#password', sandbox_user_password);
    await t.click('#btnLogin');
    await t.click('.continueButton');
});
```

If you cannot share your credentials, please create an example that demonstrates the issue using sandbox accounts and PayPal demo page.

**Comment by royalkanai** [comment]
Sent you the HTTP Authentication Credentials to the designated email address.

**Comment by AlexKamaev** [comment]
Thank you for the cooperation. We've got your credentials. I tried to run the test but it failed on the following step:

```
  120 |test('Promote with PayPal', async () => {
  121 | await t
  122 | .navigateTo('/p/asdf-stgaqaoppscore8/sponsors/new')
> 123 | .expect(Selector('.js-payment-form').visible).ok()
```

It looks like the paypal form was not loaded on the page for some reason.
I tried to run the website without testcafe and faced the same issue.

I also noticed that instead of redirecting to `/p/asdf-stgaqaoppscore8/sponsors/new`, I was redirected to `/p/asdf-stgaqaoppscore8`. Is this expected behavior?

Please fix the issue or give us instructions on how to overcome it.
Alternatively, you can share any other test that reproduces the issue.
Thank you in advance.

**Comment by rob4629** [comment]
@AlexKamaev Do you want try using url params, to switch to our Canadian site? (`/p/asdf-stgaqaoppscore8/sponsors/new?country_code=CA&lang=en-CA`).

**Comment by AlexKamaev** [comment]
@rob4629 Thank you for your hint. I was able to run a test with the Paypal authentication window. However, it looks like I faced some other issue that does not allow me to research your case further.

Here is the error I encountered:
```JS
 Verify PayPal Payments
 × Promote with PayPal

   1) The iframe in which the test is currently operating is not visible anymore.

      Browser: Chrome 99.0.4844.51 / Windows 10

         49 |const paypalLogin = async () => {
         50 |    await t
         51 |        .typeText(emailInput, 'sb-43ajzc14446759@personal.example.com', { timeout: 30000 })
         52 |        .click(nextButton)
         53 |        .typeText(passwordInput, 'k|.@S2be')
       > 54 |        .click(loginButton)
         55 |        .wait(10000)
         56 |};

```

We would need to research why I am getting this error. If you have any hints on how to overcome it, please advise.

Thank you.

**Comment by rob4629** [comment]
We see that error occasionally too (_albeit usually after completing the Paypal mini-window_). I've tried using 
```js
...
  .click(payWithPaypalButton)
  .switchToMainWindow();
```
But maybe increasing the  wait (`wait(5000)`) after clicking the paypal button, would provide a more stable result.

**Comment by rob4629** [comment]
Do you want us to create a new ticket for `The iframe in which the test is currently operating is not visible anymore.`? The only issue there is I haven't been able to consistently reproduce it

**Comment by AlexKamaev** [comment]
@rob4629 Yes, please create a separate ticket. I think we need to fix the `The iframe in which the test is currently operating is not visible anymore` error first and only after that, we can research the issue further. Thank you for your help and cooperation.

**Comment by rob4629** [comment]
I've created 6927. I copied the code from the test steps in this issue. Please use the same httpAuth credentials from this ticket

**Comment by aurelienlair** [comment]
We also have noticed this problem with PayPal guys. I wanted to underline that on our case we updated TestCafe's version but it was _also_ happening with previous versions.

**Comment by Aleksey28** [comment]
Thank you for your feedback. We don't have any updates yet. We will update this thread once we have any news.

**Comment by miherlosev** [comment]
Duplicate of https://github.com/DevExpress/testcafe/issues/7707
